patroni/postgresql.py

```diff
@@ -622,12 +658,84 @@ def reload(self):
             self.call_nowait(ACTION_ON_RELOAD)
         return ret
 
-    def restart(self):
-        self.set_state('restarting')
-        ret = self.stop(block_callbacks=True) and self.start(block_callbacks=True)
-        if ret:
-            self.call_nowait(ACTION_ON_RESTART)
+    def check_for_startup(self):
+        """Checks PostgreSQL status and returns if PostgreSQL is in the middle of startup."""
+        if self.is_starting():
+            return not self.check_startup_state_changed()
         else:
+            ready = self.pg_isready()
+            if ready == STATE_REJECT:
+                # Maybe add a ACTION_ON_FAIL here?
+                self.set_state('starting')
+                return True
+            else:
+                # Other cases will be handled by is_healthy check
+                return False
+
+    def check_startup_state_changed(self):
+        """Checks if PostgreSQL has completed starting up or failed or still starting.
+
+        Should only be called when state == 'starting'
+
+        :returns: True iff state was changed from 'starting'
+        """
+        ready = self.pg_isready()
+
+        if ready == STATE_REJECT:
+            return False
+        elif ready == STATE_NO_RESPONSE:
```

Comment:
> Alternative would be to start postmaster directly. 

It is rather simple, pg_ctl does something like:
```python
    pid = os.fork()
    if pid == 0:
        os.setsid()  # Creates a session and sets the process group ID (to avoid propagating Ctrl+C to postgres)
        # Environment must be empty, to avoid passwords leak
        os.execle("/bin/sh", "/bin/sh", "-c", 'exec "postgres" -D foobar --parameter="value" < "/dev/null" 2>&1', {})
    return pid
```